API/Pagination.md

```diff
@@ -0,0 +1,47 @@
+
+Rather than showing everything at once, we can limit the number of resources returned in a request and provide a way to fetch the rest. In other words, we can page through resources.
+
+There are different ways to enable pagination and to paginate resources, however the following paragraphs are limited to our most commonly used methods. That is not to say these are the only methods available, you are encouraged to use something else if it benefits your project.
+
+## Enabling pagination
+
+In most cases, pagination is enabled through query string. Query string lets clients pass additional parameters which are not appropriate to send in headers or the body. A URL with pagination applied might look like this: `/api/v1/countries?page=5`. In the example, we're fetching countries and requesting the 5th page of results. Query string is the preferred way to enable pagination and also the one we use.
+
+## Types
+
+There are many pagination types, and working with APIs you'll find different implementations. We'll stick to two types which are the most prevalent.
+
+### Page-based
+
+Page-based pagination divides a set of records into pages of equal size (except for the last page which can have less records than other pages). A number is assigned to every page, starting with 1.
+
+API should support client-set page size and page number params. An endpoint with pagination params applied could look like this: `/api/v1/countries?page[size]=50&page[number]=3`. In the example, the client requests the 3rd page of countries with maximum 50 records.
+
+If the client omits page size param you should set a default value. You should also set a maximum value (so clients can't request all records by providing an absurdly large number). If a client requests more records than the maximum allows, you can either clamp the value to the maximum or respond with an error, whatever you think is the right choice.
+
+When page number param is omitted, you should respond with the first page. Respond with an error if the param is lesser than 1, but if it's greater than the total number of pages then return the last page. Returning the last page instead of responding with an error prevents an edge case when pages disappear due to records being deleted but clients still requesting the missing page (which they think exists).
+
+Response should optionally include pagination metadata:
+- current page (in case it differs from the requested page)
+- total number of pages
+- total number of items
+This information is optional because it won't always be used by clients, and, since it typically requires executing an additional SQL COUNT query, it can be a costly operation.
+
+Since this type of pagination allows clients to request any page, it is usually reperesented in the UI with a paginator where you can select a page (e.g. `<- 2 3 4 5 6 ->`). If this is the kind of UI you're implementing pagination for, then page-based type is the right choice.
```

Comment:
I think it'd be useful to add pictures of UI elements appropriate for each pagination type, what do you think?

Reply:
what would be the image for infinite scroll pagination? 😄